libsolidity/codegen/CompilerUtils.cpp

```diff
@@ -338,14 +341,33 @@ void CompilerUtils::abiDecodeV2(TypePointers const& _parameterTypes, bool _fromM
 
 void CompilerUtils::zeroInitialiseMemoryArray(ArrayType const& _type)
 {
-	auto repeat = m_context.newTag();
-	m_context << repeat;
-	pushZeroValue(*_type.baseType());
-	storeInMemoryDynamic(*_type.baseType());
-	m_context << Instruction::SWAP1 << u256(1) << Instruction::SWAP1;
-	m_context << Instruction::SUB << Instruction::SWAP1;
-	m_context << Instruction::DUP2;
-	m_context.appendConditionalJumpTo(repeat);
+	if (_type.baseType()->hasSimpleZeroValueInMemory())
+	{
+		solAssert(_type.baseType()->isValueType(), "");
+		Whiskers templ(R"({
+			let size := mul(length, <element_size>)
+			calldatacopy(memptr, calldatasize(), size)
```

Comment:
Why is this using `calldatacopy` for zero initialisation?

Reply:
I think it is the cheapest way to do it.

Reply:
We can also use codecopy.

Reply:
Do you want me to change it?

Reply:
Ah I see, it is copying from after the actual `calldata` :) Perhaps a comment describing it or a helper named "zero fill memory" would be useful, which then describes this (and could be changed in the future if the VM is more restricted, etc).

Reply:
I will add a comment - we do not yet have the infrastructure to use helpers at this point. I will also change it to `codecopy`, because we are used to fetching constants from code, but accessing calldata outside of the valid range might be flagged as an error by some analysis tools.

Reply:
Please add a comment here explaining what it does and reasoning (aka "cheap").